### PR TITLE
Add getDealPublishedBy methods in DealManager interface

### PR DESCRIPTION
**Pull request by lisasgoh**
Methods are to retrieve a list of deals from particular users/tags/restaurants

### CODE REVIEW

File: src/main/java/com/google/step/datamanager/DealManager.java [hunk 1]
```diff
@@ -17,4 +18,16 @@ public Deal createDeal(
   public Deal updateDeal(Deal deal);
 
   public void deleteDeal(long id);
+
+  public List<Deal> getTrendingDeals();
```
Comment (jyx11011) [inline src/main/java/com/google/step/datamanager/DealManager.java:23]: Maybe add some comments to explain definition of trending deals

**Review by jyx11011 — DISMISSED** [review]

**Review by limli — COMMENTED** [review]
Merging this will cause the whole project to be not compilable. Maybe add some stub methods that implement these methods into DealManagerDatastore class?

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java [hunk 2]
```diff
@@ -105,4 +136,66 @@ public Deal updateDeal(Deal deal) {
...
+  private List<Deal> getDealsPublishedByFollowedRestaurantsOrUsers(
+      long userId, String fieldName, String filterAttribute) {
+    List<Deal> dealResults = new ArrayList<>();
+    List<Long> idsOfFollowedFieldName = followManager.getFollowedSomething(userId, fieldName);
+    for (Long id : idsOfFollowedFieldName) {
+      Filter propertyFilter = new FilterPredicate(filterAttribute, FilterOperator.EQUAL, id);
+      Query query = new Query("Deal").setFilter(propertyFilter);
+      PreparedQuery pq = datastore.prepare(query);
+      for (Entity entity : pq.asIterable()) {
+        dealResults.add(readDeal(entity.getKey().getId()));
```
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java:160]: make a `transformEntityToDeal` function and call it here and everywhere else that might need it

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java [hunk 3]
```diff
@@ -105,4 +136,66 @@ public Deal updateDeal(Deal deal) {
     searchManager.putDeal(deal, new ArrayList<>());
     return readDeal(deal.id);
   }
+
+  /** Retrieves deals posted by _ followed by user */
+  private List<Deal> getDealsPublishedByFollowedRestaurantsOrUsers(
```
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java:151]: can you split up this into 2 functions `getDealsPublishedByFollowedRestaurants` and `getDealsPublishedByFollowedUsers` ?

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java [hunk 4]
```diff
@@ -105,4 +136,66 @@ public Deal updateDeal(Deal deal) {
...
+      long userId, String fieldName, String filterAttribute) {
```
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java:152]: I'm also not sure what's the difference between `fieldName` and `filterAttribute` used for

File: src/main/java/com/google/step/datamanager/DealManagerDatastore.java [hunk 5]
```diff
@@ -105,4 +136,66 @@ public Deal updateDeal(Deal deal) {
...
+    for (Long id : idsOfFollowedTags) {
+      List<Long> dealIdsWithTag = dealTagManager.getDealIdsWithTag(id);
+      dealIdResults.addAll(dealIdsWithTag);
+    }
+    // Get rid of duplicate dealID (Deals with multiple tags)
+    Set<Long> dealIdSet = new HashSet<>(dealIdResults);
+    for (Long dealId : dealIdSet) {
+      dealResults.add(readDeal(dealId));
+    }
+    return dealResults;
```
Comment (limli) [inline src/main/java/com/google/step/datamanager/DealManagerDatastore.java:194]:
```suggestion
    List<Long> dealsWithoutDuplicates = new ArrayList<>(new HashSet<>(dealIdResults));
    return dealsWithoutDuplicates;
```

File: src/main/java/com/google/step/datamanager/FollowManager.java [hunk 6]
```diff
@@ -20,4 +20,6 @@
   public void unfollowTag(long followerId, long tagId);
 
   public List<Long> getFollowedTagIds(long followerId);
+
+  public List<Long> getFollowedSomething(long followerId, String fieldName);
```
Comment (limli) [inline src/main/java/com/google/step/datamanager/FollowManager.java:24]: I think this shouldn't be public so other classes can't add random fieldNames

**Pull request closed without merging**